optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/PlannerBenchmarkRunner.java

```diff
@@ -122,54 +127,134 @@ private void initBenchmarkDirectoryAndSubdirs() {
+            warmUpPopulate(futureMap, singleBenchmarkResultIndexMap, solverBenchmarkResultCycleArray, timeLeftPerCycle);
+            warmUp(futureMap, singleBenchmarkResultIndexMap, timeCycleEnd);
+            futureMap.clear();
+        }
+        warmUpEnded(originalTerminationConfigMap);
+        List<Runnable> notFinishedWarmUpList = warmUpExecutorService.shutdownNow();
+        if (!notFinishedWarmUpList.isEmpty()) {
+            throw new IllegalStateException("Impossible state: notFinishedWarmUpList is non-empty ("
+                    + notFinishedWarmUpList + ").");
+        }
+        logger.info("================================================================================");
+        logger.info("Warm up ended");
+        logger.info("================================================================================");
+    }
+
+    private void warmUpStarted(Map<SolverBenchmarkResult, TerminationConfig> originalTerminationConfigMap) {
+        for (SolverBenchmarkResult solverBenchmarkResult : plannerBenchmarkResult.getSolverBenchmarkResultList()) {
+            TerminationConfig originalTerminationConfig = solverBenchmarkResult.getSolverConfig().getTerminationConfig();
+            if (!originalTerminationConfigMap.containsKey(solverBenchmarkResult)) {
+                originalTerminationConfigMap.put(solverBenchmarkResult, originalTerminationConfig);
             }
-            logger.info("================================================================================");
-            logger.info("Warm up ended");
-            logger.info("================================================================================");
         }
     }
 
-    protected long warmUp(ProblemBenchmarkResult problemBenchmarkResult,
-            long startingTimeMillis, long warmUpTimeMillisSpentLimit, long timeLeft) {
-        for (SingleBenchmarkResult singleBenchmarkResult : problemBenchmarkResult.getSingleBenchmarkResultList()) {
-            SolverBenchmarkResult solverBenchmarkResult = singleBenchmarkResult.getSolverBenchmarkResult();
+    private void warmUpPopulate(Map<Future<SingleBenchmarkRunner>, SingleBenchmarkRunner> futureMap,
+                                ConcurrentMap<SolverBenchmarkResult, Integer> singleBenchmarkResultIndexMap,
+                                SolverBenchmarkResult[] solverBenchmarkResultArray,
+                                long timeLeftPerSolverConfig) {
+        for (SolverBenchmarkResult solverBenchmarkResult : solverBenchmarkResultArray) {
             TerminationConfig originalTerminationConfig = solverBenchmarkResult.getSolverConfig().getTerminationConfig();
             TerminationConfig tmpTerminationConfig = originalTerminationConfig == null
                     ? new TerminationConfig() : originalTerminationConfig.clone();
-            tmpTerminationConfig.shortenTimeMillisSpentLimit(timeLeft);
+            tmpTerminationConfig.overwriteTimeMillisSpentLimit(timeLeftPerSolverConfig);
```

Comment:
Why overwrite instead of shorten? If we have a warmup budget of 1000 seconds in this cycle, but we have a solver config that has a termination set to 50 seconds, I think we should just run it for 50 seconds on dataset 1, then 50 seconds on dataset 2, etc.